File: bindings/python/swift-sharing-request-bind.py
"""Async Python bindings for the swift-x-account-sharing backend."""


import json


import aiohttp


class SwiftSharingRequest:
    """Swift X Account Sharing backend client."""

    def __init__(self, url):
        """."""
        self.url = url
        self.session = aiohttp.ClientSession()

    async def __aenter__(self):
        """."""
        return self

    async def __aexit__(self, *excinfo):
        """."""
        await self.session.close()

    async def add_access_request(self, user, container, owner):
        """Add a request for container access."""
        url = "{0}/request/user/{user}/{container}".format(
            self.url,
            user,
            container
        )
        params = {
            "owner": owner,
        }

        async with self.session.post(url, params=params) as resp:
            return json.loads(await resp.text())

    async def list_made_requests(self, user):
        """List requests made by user."""
        url = "{0}/request/user/{1}".format(
            self.url,
            user
        )

        async with self.session.get(url) as resp:
            return json.loads(await resp.text())

    async def list_owned_requests(self, user):
        """List requests owned by the user."""
        url = "{0}/request/owner/{1}".format(
            self.url,
            user
        )

        async with self.session.get(url) as resp:
            return json.loads(await resp.text())

    async def list_container_requests(self, container):
        """List requests made for a container."""
        url = "{0}/request/container/{1}".format(
            self.url,
            container
        )

        async with self.session.get(url) as resp:
            return json.loads(await resp.text())

    async def share_delete_access(
            self,
            username,
            container,
            owner
    ):
        """Delete the details of an existing access request."""
        url = "{0}/request/user/{1}/{2}".format(self.url, username, container)
        params = {
            "owner": owner,
        }

        async with self.session.delete(url, params=params) as resp:
            return bool(resp.status == 200)

File: bindings/js/swift-sharing-request-bind.js
// Swift cross account container sharing API JavaScript bindings module.

class SwiftSharingRequest {
// Swift sharing request backend client.

  constructor (
      address
  ) {
      this.address = address;
  }

  async addAccessRequest(
      username,
      container,
      owner
  ) {
    // Add a request for container access.
    let url = new URL("/request/user/".concat(username, "/", container),
                      this.address);
    url.searchParams.append("owner", owner);

    let resp = fetch(
      url, {method: "POST"}
    ).then(
      (resp) => {return resp.json();}
    );
    return resp;
  }

  async listMadeRequests(
    username
  ) {
    let url = new URL("/request/user/".concat(username));
    let resp = fetch(
      url, {method: "GET"}
    ).then(
      (resp) => {return resp.json();}
    );
    return resp;
  }

  async listOwnedRequests(
    username
  ) {
    let url = new URL("/request/owner/".concat(username));
    let resp = fetch(
      url, {method: "GET"}
    ).then(
      (resp) => {return resp.json();}
    );
    return resp;
  }

  async listContainerRequests(
    container
  ) {
    let url = new URL("/request/container/".concat(container));
    let resp = fetch(
      url, {method: "GET"}
    ).then(
      (resp) => {return resp.json();}
    );
    return resp;
  }

  async shareDeleteAccess(
      username,
      container,
      owner
  ) {
    // Delete the details of an existing share action.
    let url = new URL(
    "/request/user/".concat(username, "/", container), this.address
    );
    url.searchParams.append("owner", owner);
    let deleted = fetch(
      url, {method: "DELETE"}
    ).then(
      (resp) => {
        if (resp.status == 200) {return true;}
        else {return false;}
      }
    );
    return deleted;
  }    
}

export default SwiftSharingRequest;
